File: homework_week2/task1.js
stations_list1=[
    "Songshan",
    "Nanjing Sanmin",
    "Taipei Arena",
    "Nanjing Fuxing",
    "Songjiang Nanjing",
    "Zhongsan",
    "Beimon",
    "Ximen",
    "Xiaonanmon",
    "CKS Memorial Hall",
    "Guting",
    "Taipower Building",
    "Gongguan",
    "Wanlong",
    "Jingmei",
    "Dapinglin",
    "Qizhang",
    "Xindian City Hall",
    "Xindian",
]

stations_list2=["Xiaobitan"]

function findAndPrint(messages, currentStation){
    // your code here
    var dictionary={}
    var distance=[]   

    for (let [name, message] of Object.entries(messages)) {//Object.entries就像python的dic.items()一樣，會返回key和value
        // 遍歷 stations_list1 中的每一個車站名稱
        for (let i = 0; i < stations_list1.length; i++) {
            let station = stations_list1[i];
            // 檢查 station 是否已經存在於 dictionary 的值中
            if (!Object.values(dictionary).includes(station) && message.includes(station)) {
                //Object.values就像python的dic.value()一樣，會返回value
                // 如果車站名沒有已經存在在dictionary中，就將 name 和 station 添加到 dictionary 中
                Object.assign(dictionary, { [name]: station });//相對於python中的dic.update
                // 跳出內層循環，已經找到一個符合條件的 station
                //break;
            }
        }

        for (let i = 0; i < stations_list2.length; i++) {
            let station = stations_list2[i];
            // 檢查 station 是否已經存在於 dictionary 的值中
            if (!Object.values(dictionary).includes(station) && message.includes(station)) {
                //Object.values就像python的dic.value()一樣，會返回value
                // 如果車站名沒有已經存在在dictionary中，就將 name 和 station 添加到 dictionary 中
                Object.assign(dictionary, { [name]: station });//Object.assign相對於python中的dic.update
                // 跳出內層循環，已經找到一個符合條件的 station
                //break;
            }
        }
    }

    //console.log(dictionary)

    if (stations_list1.includes(currentStation)){
        for (let location of Object.values(dictionary)){
            //console.log(location)
              if (stations_list1.includes(location)){
                   //distance.append(abs(stations_list1.index(current_station) - stations_list1.index(location)))
                   distance.push(Math.abs(stations_list1.indexOf(currentStation) - stations_list1.indexOf(location)));
                   //console.log("print")
                }
                   
              else if (stations_list2.includes(location)){
                   //distance.append(abs(stations_list1.index(current_station) - stations_list1.index("Qizhang"))+1 )
                   distance.push(Math.abs(stations_list1.indexOf(currentStation) - stations_list1.indexOf("Qizhang"))+1);
                   //console.log("print1")
                }
        }        
    }

    else if (stations_list2.includes(currentStation)){
        //console.log("print5")
        for (let location of Object.values(dictionary)){
            if (stations_list1.includes(location)){
                 //distance.append(abs(stations_list1.index(current_station) - stations_list1.index(location)))
                 distance.push(Math.abs(stations_list1.indexOf(currentStation) - stations_list1.indexOf("Qizhang"))+1);
                 //console.log("print2")
              }
                 
            else if (stations_list2.includes(location)){
                 //distance.append(abs(stations_list1.index(current_station) - stations_list1.index("Qizhang"))+1 )
                 distance.push(Math.abs(stations_list1.indexOf(currentStation) - stations_list1.indexOf(location)));
                 //console.log("print3")
              }
      } 
    }
    var min_distance_index=0;
    min_distance_index=distance.indexOf(Math.min(...distance));
    console.log(Object.keys(dictionary)[min_distance_index])
}
const messages={
"Bob":"I'm at Ximen MRT station.",
"Mary":"I have a drink near Jingmei MRT station.",
"Copper":"I just saw a concert at Taipei Arena.",
"Leslie":"I'm at home near Xiaobitan station.",
"Vivian":"I'm at Xindian station waiting for you.",
//"Vivi":"I'm at Jingmei MRT station."
};
findAndPrint(messages, "Wanlong"); // print Mary
findAndPrint(messages, "Songshan"); // print Copper
findAndPrint(messages, "Qizhang"); // print Leslie
findAndPrint(messages, "Ximen"); // print Bob
findAndPrint(messages, "Xindian City Hall"); // print Vivian

File: homework_week2/task1.py
stations_list1=[
    "Songshan",
    "Nanjing Sanmin",
    "Taipei Arena",
    "Nanjing Fuxing",
    "Songjiang Nanjing",
    "Zhongsan",
    "Beimon",
    "Ximen",
    "Xiaonanmon",
    "CKS Memorial Hall",
    "Guting",
    "Taipower Building",
    "Gongguan",
    "Wanlong",
    "Jingmei",
    "Dapinglin",
    "Qizhang",
    "Xindian City Hall",
    "Xindian",
]

stations_list2=["Xiaobitan"]

def find_and_print(messages, current_station):
    dictionary={}
    distance=[]   

    for name, message in messages.items():
        for station in stations_list1:
                if station not in dictionary.values() and station in message:
                     dictionary.update({name:station})
            #印出在主線的站牌，且若站牌在該字典沒有重複，則新增進字典
        for station in stations_list2:
                if station not in dictionary.values() and station in message:
                    dictionary.update({name:station})
            #印出在支線的站牌，且若站牌在該字典沒有重複，則新增進字典

    #for name,station in dictionary.items():#如果message中有兩人在同一站就只顯示第一人的名稱
    #     for name2,station2 in dictionary.items():
    #          if name!=name2 and station==station2:
    #               dictionary.pop(name2) #不可以用pop()否則會導致dictionary結果亂掉
    #print(dictionary)
                    

    
    #print(dictionary)
    #{'Leslie': 'Xiaobitan', 'Bob': 'Ximen', 'Mary': 'Jingmei', 'Copper': 'Taipei Arena', 'Vivian': 'Xindian', 'Vivi': 'Qizhang'}

    if current_station in stations_list1:
         for name,location in dictionary.items():
              if location in stations_list1:
                   distance.append(abs(stations_list1.index(current_station) - stations_list1.index(location)) )
                   
              elif location in stations_list2:
                   distance.append(abs(stations_list1.index(current_station) - stations_list1.index("Qizhang"))+1 )
                
    elif current_station in stations_list2:    
         for name,location in dictionary.items():
              if location in stations_list1:
                   distance.append(abs(stations_list1.index(current_station) - stations_list1.index("Qizhang"))+1 )
                   
              elif location in stations_list2:
                   distance.append(abs(stations_list1.index(current_station) - stations_list1.index(location)) )
                   
         #distance.append(abs(stations_list1.index(item) - stations_list.index(station)) + 1)
    #print(distance)#將每個人離你的距離放入distance列表中並列出每個人離你的距離
    min_distance_index=0
    min_distance_index=distance.index(min(distance))#找出最小的距離的那個人
    #print(min_distance_index)
    print(list(dictionary.keys())[min_distance_index])#印出那個人的名字



# your code here
messages={
"Leslie":"I'm at home near Xiaobitan station.",
"Bob":"I'm at Ximen MRT station.7",
"Mary":"I have a drink near Jingmei MRT station.14",
"Copper":"I just saw a concert at Taipei Arena.2",
"Vivian":"I'm at Xindian station waiting for you.18",
#"Vivi":"I'm at Xiaobitan station waiting for you.18"
}

find_and_print(messages, "Wanlong") # print Mary
find_and_print(messages, "Songshan") # print Copper
find_and_print(messages, "Qizhang") # print Leslie
find_and_print(messages, "Ximen") # print Bob
find_and_print(messages, "Xindian City Hall") # print Vivian
